lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java

```diff
@@ -156,11 +169,39 @@ public void onNext(final Identifier id) {
   }
 
   @Override
+  public <T> ConnectionFactory<T> registerConnectionFactory(
+      final Identifier connectionFactoryId,
+      final Codec<T> codec,
+      final EventHandler<Message<T>> eventHandler,
+      final LinkListener<Message<T>> linkListener,
+      final Identifier endPointId) {
+    if (isClosed.get()) {
+      throw new NetworkRuntimeException("Unable to register new ConnectionFactory to closed NetworkConnectionService");
+    }
+
+    final String id = connectionFactoryId.toString();
+    if (connFactoryMap.get(id) != null) {
+      throw new NetworkRuntimeException("ConnectionFactory " + connectionFactoryId + " was already registered.");
+    }
+
+    final NetworkConnectionFactory<T> connectionFactory = new NetworkConnectionFactory<>(
+        this, connectionFactoryId, codec, eventHandler, linkListener, endPointId);
+    nameServiceRegisteringStage.onNext(new Tuple<>(endPointId, (InetSocketAddress) transport.getLocalAddress()));
+
+    if (connFactoryMap.putIfAbsent(id, connectionFactory) != null) {
+      throw new NetworkRuntimeException("ConnectionFactory " + connectionFactoryId + " was already registered.");
+    }
+
+    return connectionFactory;
+  }
+
+  @Override
   public void unregisterConnectionFactory(final Identifier connFactoryId) {
     final String id = connFactoryId.toString();
     final ConnectionFactory  connFactory = connFactoryMap.get(id);
     if (connFactory != null) {
       final ConnectionFactory cf = connFactoryMap.remove(id);
+      nameServiceUnregisteringStage.onNext(connFactoryId);
```

Comment:
This should be `endpointId`, right?

But maybe more importantly: What happens if we have a ConnectionFactory `A` that registers with endpointId `X` and ConnectionFactory `B` that also registers with endpointId `X`? In this case, unregistering ConnectionFactory `A` will remove `X` from the name server. Even if `X` is still being used by ConnectionFactory `B`. Is this a bug?